pkg/block/fetcher.go

```diff
@@ -0,0 +1,410 @@
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+
+	cacheDir := ""
+	if dir != "" {
+		cacheDir = filepath.Join(dir, "meta-syncer")
+		if err := os.MkdirAll(cacheDir, os.ModePerm); err != nil {
+			return nil, err
+		}
+	}
+
+	return &MetaFetcher{
+		logger:      log.With(logger, "component", "block.MetaFetcher"),
+		concurrency: concurrency,
+		bkt:         bkt,
+		cacheDir:    cacheDir,
+		metrics:     newSyncMetrics(r),
+		filters:     filters,
+		cached:      map[ulid.ULID]*metadata.Meta{},
+	}, nil
+}
+
+var (
+	ErrorSyncMetaNotFound  = errors.New("meta.json not found")
+	ErrorSyncMetaCorrupted = errors.New("meta.json corrupted")
+)
+
+// loadMeta returns metadata from object storage or error.
+// It returns `ErrorSyncMetaNotFound` and `ErrorSyncMetaCorrupted` sentinel errors in those cases.
+func (s *MetaFetcher) loadMeta(ctx context.Context, id ulid.ULID) (*metadata.Meta, error) {
+	var (
+		metaFile       = path.Join(id.String(), MetaFilename)
+		cachedBlockDir = filepath.Join(s.cacheDir, id.String())
+	)
+
+	// TODO(bwplotka): If that causes problems (obj store rate limits), add longer ttl to cached items.
+	// For 1y and 100 block sources this generates ~1.5-3k HEAD RPM. AWS handles 330k RPM per prefix.
+	// TODO(bwplotka): Consider filtering by consistency delay here (can't do until compactor healthyOverride work).
+	ok, err := s.bkt.Exists(ctx, metaFile)
+	if err != nil {
+		return nil, errors.Wrapf(err, "meta.json file exists: %v", metaFile)
+	}
+	if !ok {
+		return nil, ErrorSyncMetaNotFound
+	}
+
+	if m, seen := s.cached[id]; seen {
+		return m, nil
+	}
+
+	// Best effort load from local dir.
+	if s.cacheDir != "" {
+		m, err := metadata.Read(cachedBlockDir)
+		if err == nil {
+			return m, nil
+		}
+
+		if !stderrors.Is(err, os.ErrNotExist) {
+			level.Warn(s.logger).Log("msg", "best effort read of the local meta.json failed; removing cached block dir", "dir", cachedBlockDir, "err", err)
+			if err := os.RemoveAll(cachedBlockDir); err != nil {
+				level.Warn(s.logger).Log("msg", "best effort remove of cached dir failed; ignoring", "dir", cachedBlockDir, "err", err)
+			}
+		}
+	}
+
+	level.Debug(s.logger).Log("msg", "download meta", "name", metaFile)
+
+	r, err := s.bkt.Get(ctx, metaFile)
+	if s.bkt.IsObjNotFoundErr(err) {
+		// Meta.json was deleted between bkt.Exists and here.
+		return nil, errors.Wrapf(ErrorSyncMetaNotFound, "%v ", err)
+	}
+	if err != nil {
+		return nil, errors.Wrapf(err, "get meta file: %v", metaFile)
+	}
+
+	defer runutil.CloseWithLogOnErr(s.logger, r, "close bkt meta get")
+
+	metaContent, err := ioutil.ReadAll(r)
+	if err != nil {
+		return nil, errors.Wrapf(err, "read meta file: %v", metaFile)
+	}
+
+	m := &metadata.Meta{}
+	if err := json.Unmarshal(metaContent, m); err != nil {
+		return nil, errors.Wrapf(ErrorSyncMetaCorrupted, "meta.json %v unmarshal: %v ", metaFile, err)
```

Comment:
Small nit: that space at the end seems unnecessary. The same happened on some other lines as well.